### PR TITLE
FEATURE: Add RSS feed to custom lists

### PR DESCRIPTION
**Pull request by Grubba27**
- Add list.rss.erb template to render RSS feed for custom lists. 
- Update the controller with similar logic to what `/filter` does.
-  Add a new route to serve the RSS feed
- Add tests for the new route

### CODE REVIEW

File: app/views/list/list.rss.erb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<rss version="2.0" xmlns:discourse="http://www.discourse.org/" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:dc="http://purl.org/dc/elements/1.1/">
+  <channel>
+    <% lang = SiteSetting.find_by_name('default_locale').try(:value) %>
```
Comment (Grubba27) [inline app/views/list/list.rss.erb:4]: I just copied [`list.rss.erb`](https://github.com/discourse/discourse/blob/01341f33e3f16777767d0ad80008c52d4d44d941/app/views/list/list.rss.erb#L1-L37) maybe I could use the one from core?

File: app/controllers/discourse_custom_topic_lists/lists_controller.rb [hunk 2]
```diff
@@ -3,9 +3,51 @@
...
 
     def show
       render json: success_json
     end
+
+    def lists_feed
+      discourse_expires_in 1.minute
+
+      raise Discourse::NotFound if !SiteSetting.experimental_topics_filter
+      raise Discourse::InvalidParameters if !params[:topic_list_name]
```
Comment (tgxworld) [inline app/controllers/discourse_custom_topic_lists/lists_controller.rb:16]:
```suggestion
      raise Discourse::InvalidParameters.new(:topic_list_name) if !params[:topic_list_name]
```

File: app/controllers/discourse_custom_topic_lists/lists_controller.rb [hunk 3]
```diff
@@ -3,9 +3,51 @@
...
+      topic_query_opts[:q] = list_item["query"]
+      %i[page q].each do |key|
+        if params.key?(key.to_s)
+          value = params[key]
+          raise Discourse::InvalidParameters.new(key) if !TopicQuery.validate?(key, value)
+          topic_query_opts[key] = value
+        end
+      end
+
+      user = list_target_user
```
Comment (tgxworld) [inline app/controllers/discourse_custom_topic_lists/lists_controller.rb:34]:
```suggestion
```

It doesn't seem like this line is necessary.

File: spec/requests/lists_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+RSpec.describe "Lists controller" do
```
Comment (tgxworld) [inline spec/requests/lists_controller_spec.rb:3]:
```suggestion
RSpec.describe ListsController do
```

File: config/routes.rb [hunk 5]
```diff
@@ -1,5 +1,9 @@
 # frozen_string_literal: true
 
-DiscourseCustomTopicLists::Engine.routes.draw { get "lists/:topic_list_name" => "lists#show" }
+DiscourseCustomTopicLists::Engine.routes.draw do
+  get "list/:topic_list_name",
+      to: redirect { |params, request| "lists/#{params[:topic_list_name]}" }
```
Comment (tgxworld) [inline config/routes.rb:5]: Just curious, why is this redirect necessary?

File: app/controllers/discourse_custom_topic_lists/lists_controller.rb [hunk 6]
```diff
@@ -3,9 +3,51 @@
...
+
+    def lists_feed
+      discourse_expires_in 1.minute
+
+      raise Discourse::NotFound if !SiteSetting.experimental_topics_filter
+      raise Discourse::InvalidParameters if !params[:topic_list_name]
+      topic_query_opts = { no_definitions: !SiteSetting.show_category_definitions_in_topic_lists }
+
+      list_item =
+        DiscourseCustomTopicLists::CustomTopicList.new(current_user).find_by_slug(
```
Comment (tgxworld) [inline app/controllers/discourse_custom_topic_lists/lists_controller.rb:20]:
```suggestion
        DiscourseCustomTopicLists::CustomTopicList.new(list_target_user).find_by_slug(
```

Should this be like the above?

File: app/views/lists/list.rss.erb [hunk 7]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<rss version="2.0" xmlns:discourse="http://www.discourse.org/" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:dc="http://purl.org/dc/elements/1.1/">
+  <channel>
+    <% lang = SiteSetting.find_by_name('default_locale').try(:value) %>
+    <% site_email = SiteSetting.find_by_name('contact_email').try(:value) %>
```
Comment (tgxworld) [inline app/views/lists/list.rss.erb:5]:
```suggestion
    <% lang = SiteSetting.default_locale %>
    <% site_email = SiteSetting.contact_email %>
```

I don't think we need to query the database for the site settings here.

File: plugin.rb [hunk 8]
```diff
@@ -14,28 +14,15 @@ module ::DiscourseCustomTopicLists
 end
 
 require_relative "lib/discourse_custom_topic_lists/engine"
+require_relative "lib/discourse_custom_topic_lists/custom_topic_list"
 register_asset "stylesheets/common/common.scss"
 
 after_initialize do
+  require_relative "app/controllers/discourse_custom_topic_lists/lists_controller"
```
Comment (tgxworld) [inline plugin.rb:21]: Can you see if you can update the plugin to take advantage of Rails' autoloading? https://meta.discourse.org/t/structuring-a-plugin-for-rails-autoloading/256092 

Manually requiring files should not be required.

File: plugin.rb [hunk 9]
```diff
@@ -14,28 +14,15 @@ module ::DiscourseCustomTopicLists
...
+
+  Discourse::Application.routes.append { mount ::DiscourseCustomTopicLists::Engine, at: "" }
```
Comment (tgxworld) [inline plugin.rb:23]: I'm curious why this change is necessary. The previous line to mount the engine's route in `config/routes.rb` should work.

File: spec/requests/lists_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+      },
+    )
+  end
+
+  describe "GET /list/custom-topic-list.rss" do
+    it "should return a list of custom topic lists same as `/filter`" do
+      get "/filter.json?q=category:arts-media", headers: { "ACCEPT" => "application/json" } # same as the UI would do
+      expect(response.status).to eq(200)
+      json = response.parsed_body
+      expect(json["topic_list"]["topics"].size).to eq(1)
```
Comment (tgxworld) [inline spec/requests/lists_controller_spec.rb:73]: I'm not sure if I agree with this. Per the `describe` description, we are testing the `/list/custom-topic-list.rss` endpoint so why do we need to assert for the response of the `/filter` route which this plugin does not have an effect on.

File: spec/requests/lists_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+      get "/lists/arts-and-media.rss"
+      expect(response.status).to eq(200)
+      expect(response.media_type).to eq("application/rss+xml")
+      expect(response.body).to include("Arts and Media")
+      expect(response.body).to include("## Topic with categories related to arts and media")
+      expect(response.body).to include("arts-media")
```
Comment (tgxworld) [inline spec/requests/lists_controller_spec.rb:80]: I think the tests need to be more extensive here. The template which is introduced in this PR does alot more than what we are asserting for here. Personally, I would test for something like this:

```
expect(response.body).to include(<<~RSS)
  # Full RSS output here
RSS
```

File: config/routes.rb (hunk 5 above)
Comment (Grubba27) [reply config/routes.rb:5]: While I was testing, I thought it could be an excellent addition, people might forget the `s` in `/lists/<name>`.

But after sleeping on it does not make that much sense, I'll remove it!

File: app/controllers/discourse_custom_topic_lists/lists_controller.rb [hunk 12]
```diff
@@ -3,9 +3,49 @@
...
+      @link = "#{Discourse.base_url}/lists/#{params[:topic_list_name]}"
+      @atom_link = "#{Discourse.base_url}/lists/#{params[:topic_list_name]}.rss"
+      @description = list_item["description"]
+      @topic_list = TopicQuery.new(list_target_user, topic_query_opts).list_filter
+
+      render "lists/list", formats: [:rss]
+    end
+
+    def list_target_user
+      if params[:user_id] && guardian.is_staff?
```
Comment (tgxworld) [inline app/controllers/discourse_custom_topic_lists/lists_controller.rb:44]: I know we copied this from core but that code in core was written close to a decade ago. Is there a reason we need to allow a staff user to view the RSS feed of another user? I feel like that is something unnecessary at the expense of more complexity.

If we need it, we should add a test for this logic since it is borderline security related. We never want to accidentally allow a non-staff user to view the topic list of another user.

Comment (Grubba27) [reply app/controllers/discourse_custom_topic_lists/lists_controller.rb:44]: I think it would be a "nice to have feature" if we end up having a user-based custom listing.

Currently, lists are group-based only, so I agree this is unnecessary.

I'll remove it!

File: app/controllers/discourse_custom_topic_lists/lists_controller.rb [hunk 13]
```diff
@@ -3,9 +3,41 @@
...
+
+    def lists_feed
+      discourse_expires_in 1.minute
+
+      raise Discourse::NotFound if !SiteSetting.experimental_topics_filter
+      raise Discourse::InvalidParameters.new(:topic_list_name) if !params[:topic_list_name]
+      topic_query_opts = { no_definitions: !SiteSetting.show_category_definitions_in_topic_lists }
+
+      list_item =
+        DiscourseCustomTopicLists::CustomTopicList.new(current_user).find_by_slug(
```
Comment (tgxworld) [inline app/controllers/discourse_custom_topic_lists/lists_controller.rb:20]: Sorry I don't mean to drag this on but I still don't think we are quite there yet in terms of test coverage especially when the custom topic lists uses groups as access control. We are relying on `DiscourseCustomTopicLists::CustomTopicList#lists` to ensure that the user belongs in the right groups before they can view a custom topic list. However, I do not see any test coverage for `DiscourseCustomTopicLists::CustomTopicList#lists` and the request specs added in this PR does not consider access control as well. If we do not have proper test coverage here, we are increasing our risk of a security issue in the future.

Comment (Grubba27) [reply app/controllers/discourse_custom_topic_lists/lists_controller.rb:20]: Sure! Thanks a lot for being so thorough with this PR!

I've added a few tests. If you have more suggestions, I'm super happy to hear them!

File: spec/lib/discourse_custom_topic_lists/custom_topic_list_spec.rb [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+  describe "#lists" do
+    it "returns all custom lists for admins" do
+      custom_lists = described_class.new(admin).lists
+      expect(custom_lists.size).to eq(2)
+    end
+
+    it "returns only lists that the user has access to" do
+      custom_lists = described_class.new(user).lists
+      expect(custom_lists.size).to eq(1)
```
Comment (tgxworld) [inline spec/lib/discourse_custom_topic_lists/custom_topic_list_spec.rb:52]:
```suggestion
      expect(custom_lists.size).to eq(1)
      expect(custom_lists[0]["name"]).to eq("Arts and Media")
```

Asserting for the size only is not sufficient because the method can incorrectly return the `New question` custom list for the given user and this assertion will still pass.

File: spec/lib/discourse_custom_topic_lists/custom_topic_list_spec.rb [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+      expect(custom_list["name"]).to eq("Arts and Media")
+    end
+
+    it "returns nil if the list does not exist" do
+      custom_list = described_class.new(admin).find_by_slug("non-existent")
+      expect(custom_list).to eq(nil)
+    end
+
+    it "returns nil if the user does not have access to the list" do
+      custom_list = described_class.new(user).find_by_slug("new-questions")
```
Comment (tgxworld) [inline spec/lib/discourse_custom_topic_lists/custom_topic_list_spec.rb:68]:
```suggestion
      custom_list = described_class.new(user).find_by_slug("some-long-path-for-questions")
```

I think this slug does not exist so this test is passing because the list does not exist and not because the user does not have access to the list.

File: spec/lib/discourse_custom_topic_lists/custom_topic_list_spec.rb [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+      expect(custom_list).to eq(nil)
+    end
```
Comment (tgxworld) [inline spec/lib/discourse_custom_topic_lists/custom_topic_list_spec.rb:70]:
```suggestion
    it "returns the list which a user has access to" do
      slug = "some-long-path-for-questions"
    
      custom_list = described_class.new(admin).find_by_slug(slug)
      expect(custom_list["name"]).to eq("New questions")
      
      custom_list = described_class.new(user).find_by_slug(slug)
      expect(custom_list).to eq(nil)
    end
```

**Review by tgxworld — APPROVED** [review]
Some more comments regarding the tests but the code looks good to me otherwise. Thank you for working through the feedback 🙏
